### ISSUE TITLE
Page Keep Reloading on Keyboard Press

### ISSUE THREAD

**Issue by newbieflutter** [pichillilorenzo/flutter_inappwebview#385]
I am using this version flutter_inappwebview: ^3.3.0+3. 

Below is my codes. The problem is that at some times it goes directly to  this page only http://*.*.*.*/pay.php and stopped. I want direct to go this page and send the post value which will redirect to the payement gateway page. . Secondly at times this works perfectly but when I press the keyboard example to enter credit card details the page gets reload and I can never complete the form. How do I resolve this issue of keyboard and the proper redirection ? I have even put this line `resizeToAvoidBottomInset: false, 
`
```
webView.postUrl(
                      url: 'http://*.*.*.*/pay.php', 
                      postData: Uint8List.fromList(utf8.encode("detail=1&amount=200&order_id=A123&name=thiscustomer&email=abc@abc.com&phone=12345678"))
                      );
```

```
class OnlinePaymentPage extends StatefulWidget {
  final String orderID;  
  final double totalPayment;  
  
  OnlinePaymentPage({
    Key key, 
    @required this.orderID,    
    @required this.totalPayment,   
  }) : super(key: key);
  @override
  _OnlinePaymentPageState createState() => new _OnlinePaymentPageState();
}
class _OnlinePaymentPageState extends State<OnlinePaymentPage>  {
 
  InAppWebViewController webView;
  String url = "";
  double progress = 0;
  
  @override
  void initState() {
     super.initState();
  }

  @override
  void dispose() {   
    super.dispose();
    
  }
  
  @override
  Widget build(BuildContext context) {
      return Scaffold(
          resizeToAvoidBottomInset: false, 
          appBar: AppBar(
            title: const Text('Online Payment'),
          ),
          body: Container(
            child: Column(children: <Widget>[
              Container(
                padding: EdgeInsets.all(5.0),
                child: Text(
                    "CURRENT URL\n${(url.length > 50) ? url.substring(0, 50) + "..." : url}"),
              ),
              
              Expanded(
                child: Container(
                  margin: const EdgeInsets.all(10.0),
                  
                  child: InAppWebView(
                    initialUrl: "http://*.*.*.*/pay.php",
                    initialHeaders: {},
                    initialOptions: InAppWebViewGroupOptions(
                      crossPlatform: InAppWebViewOptions(
                          debuggingEnabled: true,
                      )
                    ),
                    onWebViewCreated: (InAppWebViewController controller) {
                    webView = controller;
                    webView.postUrl(
                      url: 'http://*.*.*.*/pay.php', 
                      postData: Uint8List.fromList(utf8.encode("detail=1&amount=200&order_id=A123&name=thiscustomer&email=abc@abc.com&phone=12345678"))
                      );
                    },
                    onLoadStart: (InAppWebViewController controller, String url) {
                      setState(() {
                        this.url = url;
                      });
                    },
                    onLoadStop: (InAppWebViewController controller, String url) async {
                      if (url.startsWith("http://*.*.*.*/pay.php")) {
                        // get your token from url
                        RegExp regExp = new RegExp("access_token=(.*)");
                        String token = regExp.firstMatch(url)?.group(1);
                        print(token);

                        // or using CookieManager
                        //CookieManager cookieManager = CookieManager.instance();
                        //Cookie token = await cookieManager.getCookie(url: "https://myUrl.com/auth-response", name: "access_token");
                        //print(token.value);

                        // or using javascript to get access_token from localStorage
                        //String token = await controller.evaluateJavascript(source: "localStorage.getItem('access_token')");
                        print(token);
                      }
                    },
                    onProgressChanged: (InAppWebViewController controller, int progress) {
                      setState(() {
                        this.progress = progress / 100;
                      });
                    },
                  ),
                ),
              ),
              
          ])),
        
      );
  }
      
  }

```

**Comment by pichillilorenzo** [comment]
You can remove the `initialUrl` attribute (default will be `about:blank`). Then, on the `onWebViewCreated` event, you can call the `postUrl()` method as you are already doing.

If the page reloads, you need to update your `AndroidManifest.xml`. Add this to your main Android `<activity>`:
https://github.com/pichillilorenzo/flutter_inappwebview/blob/30102a5c1ac8ba44d2dc058f534eb7e4b0abfa04/example/android/app/src/main/AndroidManifest.xml#L56

**Comment by newbieflutter** [comment]
Hi Pichililorenzo,
                          I have followed your suggestion for the blank url I just put this `initialUrl: "", `like you suggested. But for the keyboard the problem is still the same below is my full AndroidManifest.xml file I replace the original value of android:configChanges with yours as below.
What else should I change to make it work smoothly. Thank you.

```
<application
        android:name="io.flutter.app.FlutterApplication"
        android:label="test1"
        android:usesCleartextTraffic="true"
        android:icon="@mipmap/ic_launcher">       
        <activity
            android:name=".MainActivity"
            android:launchMode="singleTop"
            android:theme="@style/LaunchTheme"
            android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale|layoutDirection|fontScale|screenLayout|density"
            android:hardwareAccelerated="true"
            android:windowSoftInputMode="adjustResize">
            <!-- Specifies an Android theme to apply to this Activity as soon as
                 the Android process has started. This theme is visible to the user
                 while the Flutter UI initializes. After that, this theme continues
                 to determine the Window background behind the Flutter UI. -->
            <meta-data
              android:name="io.flutter.embedding.android.NormalTheme"
              android:resource="@style/NormalTheme"
              />
            <!-- Displays an Android View that continues showing the launch screen
                 Drawable until Flutter paints its first frame, then this splash
                 screen fades out. A splash screen is useful to avoid any visual
                 gap between the end of Android's launch screen and the painting of
                 Flutter's first frame. -->
            <meta-data
              android:name="io.flutter.embedding.android.SplashScreenDrawable"
              android:resource="@drawable/launch_background"
              />
            <intent-filter>
                <action android:name="android.intent.action.MAIN"/>
                <category android:name="android.intent.category.LAUNCHER"/>
            </intent-filter>
        </activity>
        <!-- Don't delete the meta-data below.
             This is used by the Flutter tool to generate GeneratedPluginRegistrant.java -->
        <meta-data
            android:name="flutterEmbedding"
            android:value="2" />
    </application>
```

**Comment by newbieflutter** [comment]
Hi Pichililorenzo,
                           I tried few options like removing  this |keyboardHidden|keyboard etc but yet it does not work. What is your suggestion which could be wrong on my side ? Also I notice most of the times the page get stuck at this link http://*.*.*.*/pay.php and its empty which blank page? I suspect the its not calling the post method or the post function is not going through ? I confirm manually the link is fine and up always.

**Comment by newbieflutter** [comment]
Hi,
    Any solution to the issue which I have reported above hope you could help me on this cause I am eagerly finding for a solution.

**Comment by pichillilorenzo** [comment]
Delete the `initialUrl` attribute (not set to `""`, just remove it). Then try to call `postUrl` method on the `onLoadStop` event.

Instead, about the keyboard, could you make a screen record in order to see what is the problem? It should not reload the page. In any of my examples I tried, it never reloads.
Are you sure that what's is going on is a page reload or just a page resize??

**Comment by pichillilorenzo** [comment]
No sorry don't put `postUrl` method on the `onLoadStop` event. It should work on `onWebViewCreated` event.

**Comment by pichillilorenzo** [comment]
Also, use the chrome debugger to see if there is any javascript error on the console

**Comment by msarkrish** [comment]
@pichillilorenzo For me, webview gets reload when i'm opening keyboard by pressing text field, and while closing keyboard then also webview is reloaded. how to stop this bro?

**Comment by pichillilorenzo** [comment]
@msarkrish please post the code you are using and your AndroidManifest.xml. Also, could you make a screen record about it?? Thanks

**Comment by msarkrish** [comment]
@pichillilorenzo  bro i added code, manifest file and video

`<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.companyName.appname">
    
         <uses-permission android:name="android.permission.INTERNET" />
    
    <application
        android:name="io.flutter.app.FlutterApplication"
        android:label="Flathub"
        android:networkSecurityConfig="@xml/network_security_config"
        android:icon="@mipmap/ic_launcher">
        <activity
            android:name=".MainActivity"
            android:launchMode="singleTop"
            android:theme="@style/LaunchTheme"
            android:configChanges="orientation|keyboardHidden|keyboard|screenSize|smallestScreenSize|locale|layoutDirection|fontScale|screenLayout|density|uiMode"
            android:hardwareAccelerated="true"
            android:windowSoftInputMode="adjustResize">
            
            <meta-data
              android:name="io.flutter.embedding.android.NormalTheme"
              android:resource="@style/NormalTheme"
              />
            
            <meta-data
              android:name="io.flutter.embedding.android.SplashScreenDrawable"
              android:resource="@drawable/launch_background"
              />
            <intent-filter>
                <action android:name="android.intent.action.MAIN"/>
                <category android:name="android.intent.category.LAUNCHER"/>
            </intent-filter>
            <intent-filter>
    <action android:name="FLUTTER_NOTIFICATION_CLICK" />
    <category android:name="android.intent.category.DEFAULT" />
</intent-filter>
        </activity>
        
        <meta-data
            android:name="flutterEmbedding"
            android:value="2" />
    </application>
</manifest>


`
Code

`
     
    Widget build(BuildContext context) {
      return WillPopScope(
        onWillPop: () async {
         bool canGoBack = await webViewController.canGoBack();
        if (canGoBack == true) {
          webViewController.goBack();
        } else {
          return true;
        }
      },
      child: Scaffold(
        body: SafeArea(
          child: Column(
            children: [
              Container(
                  child: progress < 1.0
                      ? LinearProgressIndicator(value: progress)
                      : Container()),
              Expanded(
                child: InAppWebView(
                  initialOptions: InAppWebViewGroupOptions(
                      android: AndroidInAppWebViewOptions()),
                  initialUrl: 'web url',
                  onWebViewCreated: (controller) {
                    webViewController = controller;
                  },
                  onLoadError: (controller, url, code, message) {
                    setState(() {
                      controller.loadUrl(url: 'about:blank');
                      error = true;
                      print('error occured');
                      print(error);
                    });
                  },
                  onProgressChanged:
                      (InAppWebViewController controller, int progress) {
                    setState(() {
                      this.progress = progress / 100;
                    });
                  },
                ),
              ),
            ],
          ),
        ),  
       ),
      );
     }
`

Video url:
https://drive.google.com/file/d/1AptUX_ju1Iz6uyAIRiSW_Ety7_84BErS/view?usp=sharing

**Comment by pichillilorenzo** [comment]
It is just a resize and not a webpage reload! It's a reload if the website reloads, that seems not that case here.
It resizes itself because `android:configChanges` has `keyboardHidden|keyboard`. It is required if you want to be able to scroll to bottom, for example, the webview (or other non-native widgets) when the keyboard appears.

This ugly effect of widget "reloading" is caused to how Flutter manages Android native views and not in the plugin itself.
If you don't want this effect, you can use the `resizeToAvoidBottomInset: false` on your `Scaffold` widget, but in this way, you will have a little part of the webview not visible and not reachable.

**Comment by newbieflutter** [comment]
Hi,
    Here is my latest codes. As you suggested remove the intialUrl I have done that. Then next you suggested` resizeToAvoidBottomInset: false` . But the issue now is that it comes to the page here but its empty nothing. The above I got print and the result is as this CURRENT URL http://*.*.*.*/pay.php

```
I suspect it get stuck at my pay.php and it does not do the post or even loading the page completely ?


class OnlinePaymentPage extends StatefulWidget {
  final String orderID;  
  final double totalPayment;  
  
  OnlinePaymentPage({
    Key key, 
    @required this.orderID,    
    @required this.totalPayment,   
  }) : super(key: key);
  @override
  _OnlinePaymentPageState createState() => new _OnlinePaymentPageState();
}
class _OnlinePaymentPageState extends State<OnlinePaymentPage>  {
 
  InAppWebViewController webView;
  String url = "";
  double progress = 0;
  
  @override
  void initState() {
     super.initState();
  }

  @override
  void dispose() {   
    super.dispose();
    
  }
  
  @override
  Widget build(BuildContext context) {
      return Scaffold(
          resizeToAvoidBottomInset: false, 
          appBar: AppBar(
            title: const Text('Online Payment'),
          ),
          body: Container(
            child: Column(children: <Widget>[
              Container(
                padding: EdgeInsets.all(5.0),
                child: Text(
                    "CURRENT URL\n${(url.length > 50) ? url.substring(0, 50) + "..." : url}"),
              ),
              
              Expanded(
                child: Container(
                  margin: const EdgeInsets.all(10.0),
                  
                  child: InAppWebView(
                    initialHeaders: {},
                    initialOptions: InAppWebViewGroupOptions(
                      crossPlatform: InAppWebViewOptions(
                          debuggingEnabled: true,
                      )
                    ),
                    onWebViewCreated: (InAppWebViewController controller) {
                    webView = controller;
                    webView.postUrl(
                      url: 'http://*.*.*.*/pay.php', 
                      postData: Uint8List.fromList(utf8.encode("detail=1&amount=200&order_id=A123&name=thiscustomer&email=abc@abc.com&phone=12345678"))
                      );
                    },
                    onLoadStart: (InAppWebViewController controller, String url) {
                      setState(() {
                        this.url = url;
                      });
                    },
                    onLoadStop: (InAppWebViewController controller, String url) async {
                      if (url.startsWith("http://*.*.*.*/pay.php")) {
                        // get your token from url
                        RegExp regExp = new RegExp("access_token=(.*)");
                        String token = regExp.firstMatch(url)?.group(1);
                        print(token);

                        // or using CookieManager
                        //CookieManager cookieManager = CookieManager.instance();
                        //Cookie token = await cookieManager.getCookie(url: "https://myUrl.com/auth-response", name: "access_token");
                        //print(token.value);

                        // or using javascript to get access_token from localStorage
                        //String token = await controller.evaluateJavascript(source: "localStorage.getItem('access_token')");
                        print(token);
                      }
                    },
                    onProgressChanged: (InAppWebViewController controller, int progress) {
                      setState(() {
                        this.progress = progress / 100;
                      });
                    },
                  ),
                ),
              ),
              
          ])),
        
      );
  }
      
  }
```

**Comment by pichillilorenzo** [comment]
Ok, then probably, the problem is related to the URL scheme you are using.
If you use`HTTP` and not `HTTPS`, then you need to add support to it (see [Support HTTP request](https://github.com/pichillilorenzo/flutter_inappwebview#important-note-for-android) in Android Section).

Check this answer: https://stackoverflow.com/questions/45940861/android-8-cleartext-http-traffic-not-permitted/50834600#50834600.
As he says, if none works, the last option is to set  the attribute `android:usesCleartextTraffic="true"` in your `AndroidManifest.xml` (`<application>` tag):
```xml
<application
  ...
  android:usesCleartextTraffic="true"
  ...>
  ...
</application>
```

**Comment by newbieflutter** [comment]
Hi,
    I have tried you suggestion here android:usesCleartextTraffic="true" but no luck it still the same white empty white page.

**Comment by pichillilorenzo** [comment]
Then I don't know how to help you. If you can share your URL, I could try on my own, otherwise you need to debug yourself the problem, probably going into the Android native side code.

**Comment by newbieflutter** [comment]
Hi,
    How can I privately send you the url?

**Comment by pichillilorenzo** [comment]
You can reach me at pichillilorenzo@gmail.com

**Comment by newbieflutter** [comment]
Hi,
    Thank you I have sent an email with all the details. Thank you.

**Comment by msarkrish** [comment]
@pichillilorenzo  Bro, is there any other way to avoid page blink when keyboard is open and close, without using resizeToAvoidBottomInset:false in Scaffold.

**Comment by pichillilorenzo** [comment]
@msarkrish unfortunately no, we can only wait the flutter team to work on a better implementation of Android native view.
I know that they are working on it. I hope they will release it as soon as possible.

**Comment by pichillilorenzo** [comment]
@newbieflutter I didn't receive it yet!

**Comment by newbieflutter** [comment]
Hi,
   I have resend see if you have received it now ? Thank you.

**Comment by newbieflutter** [comment]
Hi,
    Any updates thank you.

**Comment by pichillilorenzo** [comment]
I think I can work on it tomorrow! I will let you know as soon as possible

**Comment by newbieflutter** [comment]
Hi Lorenzo,
                 Thank you for the updates.

**Comment by ycv005** [comment]
I am facing this on **particular** mobile phone that has 6 inch screen. @pichillilorenzo Any Idea why? I have already used following solution
1) `android:usesCleartextTraffic="true"`
2) Tried removing and adding back `KeyboardHidden|Keyboard`
3) Used Scaffold with and without `resizeToAvoidBottomInset: `.
4) Now Using a Container over InAppWebView class instead of Scaffold with resizeToAvoidBottomInset.

Also when it blink I am getting following log-
```
E/InputMethodManager(17883): b/117267690: Failed to get fallback IMM with expected displayId=426 actual IMM#displayId=0 view=com.pichillilorenzo.flutter_inappwebview.InAppWebView.InAppWebView{8f2339f VFED..CL. ......ID 0,0-1080,2015}
W/ContentCatcher(17883): Failed to notify a WebView
```

**Comment by pichillilorenzo** [comment]
For Android keyboards problems, they are fixed with the new version 5.0.0. Try the latest version `5.0.5+3` and use the Android-specific webview option: `useHybridComposition: true`.
